### PR TITLE
INT-3795: TCP: Combo Failover and Cached CFs

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3795

Previously, a `CachingClientConnectionFactory` could wrap `FailoverClientConnectionFactory` instances.

It was not possible to wrap `CachingClientConnectionFactory` instances in a `FailoverClientConnectionFactory`.

The failover logic only failed over when an `IOException` occurred, whereas the CCCF wrapped the IOException
in a `MessagingException`.
- Catch all exceptions in the failover logic
- Add real TCP test cases for both scenarios
  - test failover when a connection is closed
  - test failover when the first factory cannot connect at all

### CODE REVIEW

**Comment by artembilan** [comment]
Gary, I won't mind to merge - everything looks good, but pay attention to the Travis result.
We already have enough "unresolvable" sporadic failures and introducing one more would decrease the paradise in our lives :smile: .

Passes for me, BTW, as on Windows, as well as on Mac...

**Comment by garyrussell** [comment]
Added a travis debug commit (temporary)

**Comment by artembilan** [comment]
Gary, looks like Travis has done his DEBUG work already.
And what is funny exactly with the same failure!
So we are lucky here :smile:

**Comment by artembilan** [comment]
H-m-m.. It is marked like passed now...
Any ideas ?

**Comment by garyrussell** [comment]
I am debugging

**Comment by garyrussell** [comment]
Pushed.

File: .travis.yml [hunk 1]
```diff
@@ -7,5 +7,5 @@ services:
   - redis-server
 env:
   - TERM=dumb SI_FATAL_WHEN_NO_BEANFACTORY=true
-#script:
-#  - ./gradlew build --parallel
+script:
+  - ./gradlew :spring-integration-ip:build -d -Dtest.single=FailoverClientConnectionFactoryTests
```
Comment (artembilan) [inline .travis.yml]: This is wrong change for the whole project.
Fixing (reverting) on merge...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/0ce5f12342a1fb6001359dbc50d3fb788b4ce89d

**Pull request closed without merging**